### DISCUSSION TITLE
几个问题：

### DISCUSSION THREAD

**Discussion by Ice-Hazymoon** [koishijs/koishi#1017] [Q&A]
before-attach-user 事件是什么时候被调用

我创建了一个插件 A，增加了一个 attach-user 事件

```js
ctx.before('attach-user', (session) => {
    console.log('before attach-user');
})
```

插件 B，监听 message 事件

```js
ctx.on('message', async (session) => {
    const databaseUser = await ctx.database.getUser('onebot', session.userId);
    const sessionUser = await session.getUser(session.userId);
    console.log('plugin B')
})
```

收到消息后的输出

```
plugin B
before attach-user
```

我对文档的理解是应该先输出 `before attach-user`，再输出 `plugin B`

![image](https://user-images.githubusercontent.com/25239977/223564594-8c021110-8f4e-45df-92c9-a6c74ec2ffd3.png)

---

## 其他

我应该如何实现在处理所有事件之前（例如用户消息，加群退群），查询数据库判断用户是否有使用权限，如果有权限，则继续后续逻辑和插件，如果无权限，则返回“无权限”的消息提示，停止后续的处理。

**Comment by MaikoTan** [comment]
> 我应该如何实现在处理所有事件之前（例如用户消息，加群退群），查询数据库判断用户是否有使用权限，如果有权限，则继续后续逻辑和插件，如果无权限，则返回“无权限”的消息提示，停止后续的处理。

If you are trying to add this feature to your plugins (I mean the plugins you written by yourself) only, it is very straightforward to just add a check in every event listener before doing your own processing.

If you are trying to block some people from using others' plugins, you could add a filter which is on the top of the configuration page of every plugin.

------

Re: the order between `before-attach-user` event and `message` event

I think you have misunderstanding here, when you are trying to invoke the `ctx.database.getUser()` or `session.getUser()` functions, it would not trigger the `before-attach-user` and `attach-user` event. Instead, the `before-attach-user` event is manually triggerred by Koishi itself when Koishi received a message that could be parsed as a command, BEFORE we fetching the data of the current user, we trigger the `before-attach-user`.

Also, this means you could trigger the `before-attach-user` event by yourself. Just like this:

```diff
ctx.on('message', async (session) => {
+   ctx.emit('before-attach-user', session, userFields);
    const databaseUser = await ctx.database.getUser('onebot', session.userId);
    const sessionUser = await session.getUser(session.userId);
    console.log('plugin B')
})
```

**Comment by Ice-Hazymoon** [comment]
> If you are trying to add this feature to your plugins (I mean the plugins you written by yourself) only, it is very straightforward to just add a check in every event listener before doing your own processing.

Like this?

```js
ctx.on('message', async (session) => {
  const checkAuthority = await ctx.database.xxxxx
  if (checkAuthority) // handle plugin
  else return "No authority"
})
```

**Reply by MaikoTan** [reply]
Yep.

**Comment by shigma** [comment]
before-attach-user 属于中间件工作流，不会影响其他事件。因为只有中间件里才存在「Koishi 试图从数据库获取频道 / 用户信息前触发」的逻辑。

> 我应该如何实现在处理所有事件之前（例如用户消息，加群退群），查询数据库判断用户是否有使用权限，如果有权限，则继续后续逻辑和插件，如果无权限，则返回“无权限”的消息提示，停止后续的处理。

你需要手动处理每一个相关的事件。一个完整的 Koishi 项目通常有上百个事件，显然绝大部分事件都与这个 before-attach-user 无关。

**Reply by Ice-Hazymoon** [reply]
感谢回复，还有个问题，koishi 后续是否有自定义过滤器的计划

**Reply by shigma** [reply]
什么是自定义过滤器……？

**Reply by Ice-Hazymoon** [reply]
@shigma 过滤器中增加用户表某个字段的处理，例如 user_foo = bar 的时候处理插件，否则不处理

![image](https://user-images.githubusercontent.com/25239977/223716778-69a0733f-40c7-4a57-8027-8bd7a7c92e4e.png)

**Reply by shigma** [reply]
哦哦明白了，这个应该是可以做的，但我目前没有见到更多的场景不方便开展设计。你可以把你的场景分享一下。

**Reply by Ice-Hazymoon** [reply]
我暂时还真想不出什么例子，但我认为这个特性应该会非常有意义，可以避免代码中出现多次重复的逻辑。

举个栗子：if user_status = 1 则可以让用户使用插件，如果等于 0，则不返回任何消息。（我这只是瞎举的例子，目前想不到有更好的例子。）

如果我有许多自己开发的插件，则在每次收到消息时都用代码判断是否继续处理很繁琐，直接在过滤器中会简单很多。

**Reply by shigma** [reply]
暂时想不出来没关系，等需求出现的时候我们再进行讨论就好。
